ddl/rollingback.go

```diff
@@ -0,0 +1,156 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddl
+
+import (
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/meta"
+	log "github.com/sirupsen/logrus"
+)
+
+func convertAddIdxJob2RollbackJob(t *meta.Meta, job *model.Job, tblInfo *model.TableInfo, indexInfo *model.IndexInfo, err error) (int64, error) {
+	job.State = model.JobStateRollingback
+	// the second args will be used in onDropIndex.
+	job.Args = []interface{}{indexInfo.Name, getPartitionIDs(tblInfo)}
+	// If add index job rollbacks in write reorganization state, its need to delete all keys which has been added.
+	// Its work is the same as drop index job do.
+	// The write reorganization state in add index job that likes write only state in drop index job.
+	// So the next state is delete only state.
+	originalState := indexInfo.State
+	indexInfo.State = model.StateDeleteOnly
+	job.SchemaState = model.StateDeleteOnly
+	ver, err1 := updateVersionAndTableInfo(t, job, tblInfo, originalState != indexInfo.State)
+	if err1 != nil {
+		return ver, errors.Trace(err1)
+	}
+
+	if kv.ErrKeyExists.Equal(err) {
+		return ver, kv.ErrKeyExists.GenWithStack("Duplicate for key %s", indexInfo.Name.O)
+	}
+
+	return ver, errors.Trace(err)
+}
+
+// convertNotStartAddIdxJob2RollbackJob converts the add index job that are not started workers to rollingbackJob,
+// to rollback add index operations. job.SnapshotVer == 0 indicates the workers are not started.
+func convertNotStartAddIdxJob2RollbackJob(t *meta.Meta, job *model.Job, occuredErr error) (ver int64, err error) {
+	schemaID := job.SchemaID
+	tblInfo, err := getTableInfo(t, job, schemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	var (
+		unique      bool
+		indexName   model.CIStr
+		idxColNames []*ast.IndexColName
+		indexOption *ast.IndexOption
+	)
+	err = job.DecodeArgs(&unique, &indexName, &idxColNames, &indexOption)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	indexInfo := findIndexByName(indexName.L, tblInfo.Indices)
+	if indexInfo == nil {
+		job.State = model.JobStateCancelled
+		return ver, errCancelledDDLJob
+	}
+	return convertAddIdxJob2RollbackJob(t, job, tblInfo, indexInfo, occuredErr)
+}
+
+func rollingbackAddColumn(t *meta.Meta, job *model.Job) (ver int64, err error) {
+	job.State = model.JobStateRollingback
+	col := &model.ColumnInfo{}
+	pos := &ast.ColumnPosition{}
+	offset := 0
```

Comment:
Should offet be math.MaxUint64 ?

Reply:
I don't think it is necessary. The initial offset value is the same with onAddColumn.

Reply:
Actually, this param is useless but just decode it.